python3/bin/perfmon

```diff
@@ -0,0 +1,1620 @@
+                provenance_element = xmldoc.createElement("configured_on")
+                provenance_element.setAttribute("class", "SR")
+                provenance_element.setAttribute("uuid", sruuid)
+                vn.appendChild(provenance_element)
+                var = Variable(vn, self.alarm_create, self.get_default_variable_config)
+                variable_names.add(var.name)
+                append_var = True
+                vars_with_same_name = [
+                    v for v in self.secondary_variables if v.name == var.name
+                ]
+                for v in vars_with_same_name:
+                    # this list should be 0 or 1 long!
+                    # only replace variable in self.secondary_variables if its config has changed.
+                    # This way we don't reset its state
+                    if variable_configs_differ(var, v):
+                        print_debug(
+                            "Removing existing secondary variable to replace with new: %s"
+                            % v.name
+                        )
+                        self.secondary_variables.remove(v)
+                    else:
+                        print_debug(
+                            "Found existing secondary variable with same config: %s"
+                            % v.name
+                        )
+                        append_var = False
+                if append_var:
+                    print_debug(
+                        "Adding %s to set of secondary variables for host UUID=%s"
+                        % (var.name, self.uuid)
+                    )
+                    self.secondary_variables.add(var)
+
+        # Now that we have read all the xml items,
+        # delete any old variables that do not appear in the new variable_nodes
+        print_debug(
+            "Going to delete any secondary_variables not in %s" % variable_names
+        )
+        variables_to_remove = [
+            v for v in self.secondary_variables if v.name not in variable_names
+        ]
+        for v in variables_to_remove:
+            print_debug(
+                "Deleting %s from set of secondary variables for UUID=%s"
+                % (v.name, self.uuid)
+            )
+            self.secondary_variables.remove(v)
+
+
+all_xmlconfigs = {}
+sruuids_by_hostuuid = (
+    {}
+)  # Maps host uuid to a set of the uuids of the host's SRs that have other-config:perfmon
+
+
+def update_all_xmlconfigs(session):
+    """Update all_xmlconfigs, a global dictionary that maps any uuid
+    (SR, host or VM) to the xml config string in other-config:perfmon keys
+    and update sruuids_by_hostuuid which together with all_xmlconfigs allows
+    lookup of the other-config:perfmon xml of the SRs connected to a host"""
+    # `all_xmlconfigs` and `sruuids_by_hostuuid` are updated by clear() and update()
+    # pylint: disable=global-variable-not-assigned
+    global all_xmlconfigs
```

Comment:
It would be great if these mutable global variables can be eliminated.

Reply:
Also tracked by CP-48628